legacy/firmware/config.c

```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
   return sectrue == ret;
 }
 
-const uint8_t *session_getSessionId(void) {
-  if (!sessionIdCached) {
-    random_buffer(sessionId, 32);
+uint8_t session_findLRU(void) {
+  uint8_t lowest_index = MAX_SESSIONS_COUNT;
+  uint32_t lowest = sessionLRUCounter;
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage < lowest) {
+      lowest = sessionsCache[i].usage;
+      lowest_index = i;
+    }
+  }
+  // TODO: abort if lowest_index == MAX_SESSIONS_COUNT?
+  // TODO: it should never happen, so it is more of "internal" check
+  return lowest_index;
+}
+
+uint8_t session_findSession(const uint8_t *sessionId) {
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage != 0) {
+      if (memcmp(sessionsCache[i].id, sessionId, 32) == 0) {  // session found
+        return i;
+      }
+    }
+  }
+  return MAX_SESSIONS_COUNT;
+}
+
+uint8_t session_findEmpty(void) {
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage == 0) {
+      return i;
+    }
+  }
+  return MAX_SESSIONS_COUNT;
+}
+
+uint8_t *session_startSession(const uint8_t *received_session_id) {
+  int session_index = MAX_SESSIONS_COUNT;
+  if (received_session_id != NULL) {
+    session_index = session_findSession(received_session_id);
   }
-  sessionIdCached = sectrue;
-  return sessionId;
+  if (session_index != MAX_SESSIONS_COUNT) {
+    // session found in currently stored sessions
+    sessionLRUCounter++;
+    sessionsCache[session_index].usage = sessionLRUCounter;
+    activeSessionCache = sessionsCache + session_index;
+    return activeSessionCache->id;
+  } else {
+    // if not found or session_id not provided try to find an empty one
+    session_index = session_findEmpty();
```

Comment:
Do we need `session_findEmpty`? If you call `session_findLRU` and there is an empty session, then it will return the empty session won't it?

Reply:
Good catch. That should work as far as there is `<=` in `session_findLRU` as you also suggested. I will also add:

```
if (sessionsCache[i].usage == 0) {
  return i;
}
```

as a small optimalization.

Reply:
I am wondering, should we rename `session_findLRU` to reflect the fact it finds an empty session as well? It is a LRU item as well after all, so maybe not worth it?

Suggestions:
- session_findIndex
- session_findPosition
- session_nextCacheIndex

Reply:
Well I don't like the name of the function either, but mainly because of the abbreviation. I would have called it `session_find_least_recent()`, but I can accept the abbreviation, since it's only a local helper function and the comment explains what it does.

In any case I think the function name clearly includes finding an unused session. Any session which has not been used previously can be considered least recent.

Reply:
I agree. Let's go for `session_find_least_recent`. Actually, to stick to the notation let's name it `session_findLeastRecent`.

Reply:
Great!